test/unit/mallctl.c

```diff
@@ -853,6 +853,66 @@ TEST_BEGIN(test_hooks_exhaustion) {
 }
 TEST_END
 
+TEST_BEGIN(test_fragmentation) {
+	void *p = mallocx(1, 0);
+	void *q = mallocx(1, 0);
+	const void *in[2] = {p, q};
+	size_t in_sz = sizeof(const void *) * 2;
+	size_t out[6] = {-1, -1, -1, -1, -1, -1};
+	size_t out_sz = sizeof(size_t) * 6;
+	size_t out_ref[] = {-1, -1, -1, -1, -1, -1};
+	size_t out_sz_ref = sizeof(size_t) * 6;
+
+	/* Test invalid argument(s) errors */
+#define TEST_FRAG_EINVAL(a, b, c, d, why_inval) do {			\
+	assert_d_eq(mallctl("experimental.fragmentation", a, b, c, d),	\
+	    EINVAL, "Should fail when " why_inval);			\
+	assert_zu_eq(out_sz, out_sz_ref,				\
+	    "Output size touched when given invalid arguments");	\
+	assert_d_eq(memcmp(out, out_ref, out_sz_ref), 0,		\
+	    "Output content touched when given invalid arguments");	\
+} while (0)
+	TEST_FRAG_EINVAL(NULL, &out_sz, in, in_sz, "old is NULL");
+	TEST_FRAG_EINVAL(out, NULL, in, in_sz, "oldlenp is NULL");
+	TEST_FRAG_EINVAL(out, &out_sz, NULL, in_sz, "newp is NULL");
+	TEST_FRAG_EINVAL(out, &out_sz, in, 0, "newlen is zero");
+	in_sz -= 1;
+	TEST_FRAG_EINVAL(out, &out_sz, in, in_sz,
+	    "newlen is not an exact multiple");
+	in_sz += 1;
+	out_sz_ref = out_sz -= 2 * sizeof(size_t);
+	TEST_FRAG_EINVAL(out, &out_sz, in, in_sz,
+	    "*oldlenp is not an exact multiple");
+	out_sz_ref = out_sz += 2 * sizeof(size_t);
+	in_sz -= sizeof(const void *);
+	TEST_FRAG_EINVAL(out, &out_sz, in, in_sz,
+	    "*oldlenp and newlen do not match");
+	in_sz += sizeof(const void *);
+#undef TEST_FRAG_EINVAL
+
+	/* Examine output for valid call */
+	out_sz_ref = out_sz /= 2;
+	in_sz /= 2;
+	assert_d_eq(mallctl(
+			"experimental.fragmentation", out, &out_sz, in, in_sz),
```

Comment:
nit: spacing is off. Let's always do 4 spaces for second line, regardless of the nested calls.